### ISSUE TITLE
Softmax Shape

### ISSUE THREAD

**Issue by prakruti-10** [tensorflow/magenta#1474]
The softmax shape defined in _`extend_event_sequences`_ in _`music/encoder_decoder.py`_ is ambiguous. What does _`beam_size`_ refer to?
As per my understanding the softmax size is as following:
**Case 1**: Multiple sets of classes exist
In this case, softmax shape is 4-D with the dimensions as follows:
[ num_of_different_classes X num_of_event_sequences X event_in_event_sequence X softmax_size] 
where softmax_size = number of distinct classes in each case.

**Case 2**: Only one set of class
In this case, softmax shape is 3-D with the dimensions as follows:
[ num_of_event_sequences X event_in_event_sequence X softmax_size] 
where softmax_size = number of distinct classes in the set.

**Comment by twisha96** [comment]
IMHO, your explanation is correct. Beam size is just a name given to the dimension having number of event sequences. This nomenclature makes sense as the number of event sequences would be equal to the beam size of beam search.
Hence, _**[beam_size, event_num, softmax_size]**_ is equivalent to _**[ num_of_event_sequences, event_in_event_sequence, softmax_size]**_
@cghawthorne @iansimon: Can you please confirm this? Thanks in advance.

**Comment by iansimon** [comment]
The _beam_size_ parameter is the number of highest-likelihood sequences maintained during beam search.  At each generation step, _branch_factor_ next samples are taken for each sequence in the beam, and then all but the top _beam_size_ sequences (across all branches) are discarded.

Unless you're doing some kind of conditional generation (e.g. with Improv RNN), you probably want to use a beam size and branch factor of 1.
